### DISCUSSION TITLE
Swap assignee/candidateUsers/groups

### DISCUSSION THREAD

**Discussion by kstan79** [bpmnServer/bpmn-server#139] [Ideas]
according #131, 
> Use new feature of candidateUsers by setting item.candidateUsers=[supervisor.userid]

I would like to try up. however there is no documentation for update `item`.

**Comment by kstan79** [comment]
i solve by using assign() as  your example at [userassignment](https://github.com/bpmnServer/bpmn-server/blob/master/docs/userAssignment.md)
```
    response = await server.engine.assign({id: response.id , "items.elementId": 'task_Buy' },newData,userId,assignment);
```

**Comment by kstan79** [comment]
but I would like to higlight the newData override the entire `instance's data`, i suggest don't do it cause it will mess up the traceability. since this api use for assign, i think shall focus in scope of that userTask

**Comment by ralphhanna** [comment]
This is what is happening in my new release:
```js
>  let response = await api.engine.start('Vacation Request', { reason: 'I like it', type: 'Vacation' }, user1);
---------------- started -----------data: { reason: 'I like it', type: 'Vacation', requester: 'user1' }
>  response = await api.engine.assign({ id: response.id, "items.elementId": 'Request' }, { duration: '10 days' }, {},user1);
---------------- Assigned -----------data: {
  reason: 'I like it',
  type: 'Vacation',
  requester: 'user1',
  duration: '10 days'
}
---------------- Trying with user1 -----------
>    response = await api.engine.invoke({ id: response.id, "items.elementId": 'Request' }, {duration: '20 days'}, user1);

---------------- Requested -----------data: {
  reason: 'I like it',
  type: 'Vacation',
  requester: 'user1',
  duration: '20 days'
}
```
is not overwriting any data, is that what you mean?

Thanks

**Comment by kstan79** [comment]
ah, your sample override your initial data. this enforcement is less appropriate same like #135.  

Base on my understanding (correct me if i'm wrong), `assign` design for update specific user task actor(s) and others properties like due date

the generic sequence as below:
1. new workflow trigger with initial `data` =`{date:'2000-01-01', description: 'holiday with family'  },`
2. the workflow stop after initiate user task, at this moment `assignee=actor1` in bpmn definition
3. after workflow executed, hr notice the `actor1` take leave, she use `assign()` to swap assignee to `actor2`, with description `user on leave`
4. after sometimes, notice `actor2` not available, it swap again to `actor3`, with description `actor2 on-leave too`
5. now, 3rd actor wish to approve/reject task, but the data he seen been polluted. he saw
`{date:'2000-01-01', description: 'actor2 onleave too'  }`

6. no matter how, he decided to reject this leave cause he not understand, so he gave data:
`{description: 'why actor2 onleave you actor1 need onleave too', decision: 'reject'}`
7. we have no place to trace the history, and what we can see at final will be:
`{date:'2000-01-01', description: 'why actor2 onleave you actor1 need onleave too' , decision: 'reject' }`

IMHO, below is important:
1. `assign()` shall not trigger and changes of data (no matter root level, or item level), it use for assign property of usertask only
2. `invoke()` is process of making decision, the actor shall supply own data and it need to save into the user task. (it does not automatically update `root's data`)
3. in `invoke()` codes, developer can make own decision to merge the data from `item's data` to `root's data`, I recommend:
  a. `diagram way`: define  delegates / listener in bpmn, the rest let source code decide
  b. `option way`: additional `{copyToData:true}` at `assign's option`
  
in fact, this ideal is not come from my own, historical bpmn  engine I use before also isolate data by individual task, and manage variable separately in bpmn. However it too complex and too confusing. I just propose simplified way

**Reply by ralphhanna** [reply]
### Security Rules for new Release 2.0.0

### Security Rules are by-passed if
 a.  User is and Admin or System
 OR b.  .env REQUIRE_AUTHENTICATION === 'false' || process.env.ENFORCE_SECURITY
  
### User can View/Assign Item if:
  1.  Item has no restrictions if (`Assignee`==null and `CandidateUsers` == null and `CandidateGroups`==null)
  2.  If user is has a group indicated in `item.CandidateGroups`
 
### User can Execute Item
If user is the `Assignee`

  
### Scenarios:
  1.  User is Assignee
       Can Invoke/Complete Task
      Can Assign to another User
  2.  User is not Assignee but Can View (see rules above)
       Can Assign Task to either another user or self (Take Task)
       Then User can Invoke Task
 
### Notes
- We may need to utilize Optimistic Lock pattern here, keep track of version # of the Instance and on update
- Warn the User that instance information has changed since last seen

**Comment by ralphhanna** [comment]
Very interesting... please let me think about this abit ,I am converting it to a conversation

**Comment by ralphhanna** [comment]
How does that look:
```js
>  let response = await api.engine.start('Vacation Request', { reason: 'I like it', type: 'Vacation' }, user1);
data: {
  reason: 'I like it',
  type: 'Vacation',
  requester: 'user1',
  _history: {[
	{date: somedatehere,	user: user1,
		data: 
			reason: 'I like it',
			type: 'Vacation',
			requester: 'user1',
	}
	]
  }
}

---------------- started -----------data: { reason: 'I like it', type: 'Vacation', requester: 'user1' }
>  response = await api.engine.assign({ id: response.id, "items.elementId": 'Request' }, { duration: '10 days' }, {},user1);
---------------- Assigned -----------
data: {
  reason: 'I like it',
  type: 'Vacation',
  requester: 'user1',
  duration: '10 days'
  _history: {[
	{date: somedatehere,	user: user1,
		data: 
			reason: 'I like it',
			type: 'Vacation',
			requester: 'user1',
	},
	{date: bitlater,	user: user1,
		data: 
			duration: '10 days'
	}
	]
  }
}
---------------- Trying with user1 -----------
>    response = await api.engine.invoke({ id: response.id, "items.elementId": 'Request' }, {duration: '20 days'}, user1);

---------------- Requested -----------data: {
  reason: 'I like it',
  type: 'Vacation',
  requester: 'user1',
  duration: '20 days'
  _history: {[
	{date: somedatehere,	user: user1,
		data: 
			reason: 'I like it',
			type: 'Vacation',
			requester: 'user1',
	},
	{date: bitlater,	user: user1,
		data: 
			duration: '10 days'
	},
	{date: daylater,	user: user1,
		data: 
			duration: '20 days'
	}
	]
  }
}
```

**Reply by kstan79** [reply]
my opinion:
1. you added history,  it is good and easily to trace modificatono history, but:
   a. instance `data` become much large
   b. we cannot find specific data from usertask. Cause the invoke's data not located in user task. Example after invoke I wish to send notification to requester which actor invoke the usertask, and what invoke's `data` they send. obtain the invoke's `data` become tedious
2. the data still been polluted. I know your intention is make it less effort to modify the data, however combine data not always suitable. Imagine a process reach a stage require 3 userTask, run parallelly using parallel gateway. the data been override by random sequence. I still discourage update instance's `data` automatically. developer shall make decision and define override logic at suitable place which we don't know

**Comment by ralphhanna** [comment]
What is the alternative?

**Comment by ralphhanna** [comment]
Regarding your no 2
We now support concurrency, so this will not happen

**Reply by kstan79** [reply]
ops, ok ok. sorry make you stress

**Comment by kstan79** [comment]
as below image, all invoke data stay in task context only. every user task could have different kind of shape, it shouldnt update to instance's data automatically cause it will easily create confusion
![suggestion](https://github.com/bpmnServer/bpmn-server/assets/1889979/9b9a3604-c1a3-484d-b4f0-63965e421425)

**Reply by kstan79** [reply]
I borrow some explanation of variables from `kogito`:

> Variables in Kogito processes store data that is used during runtime. The Kogito BPMN modeler supports the following types of variables:

> Process variables: Variables that are visible within a specific process instance. Process variables are initialized at process creation and destroyed on process completion. 

You can assume Process variables is `instance's data`

> Local variables: Variables that are visible within a specific process component, such as a task. Local variables are initialized when the element context is initialized (when the execution workflow enters the node and execution of the onEntry action has finished, if applicable). Local variables are destroyed when the element context is destroyed (when the execution workflow leaves the element).

You can assume local variables is `task's data`, during invoke user fill in the `local variables`. In kogito, every usertask may predefine different local variables. 

There is some different here which I would like to suggest is keep the local variables into task instead of destroy it.  Just for your info kogito delete the whole instance from database after process completed, i believe the intension is maximize the database performance. I would suggest daily/weekly/(or monthly) archieve into another collection which can achieve similar result

**Comment by ralphhanna** [comment]
Are you aware that we support item.vars
That is local to the item and persistent 

Are suggesting that invoke should not take data?

**Reply by ralphhanna** [reply]
Local variables are available in item.vars and code can move/copy to instance data

However, at this time item.vars are visible to the user for input or display

**Reply by kstan79** [reply]
no, i mean invoke shall capture feedback information from actor.
What I see is it only merge into instance `data`, no store at both `item's data` or `item's vars`. I use code similar like below 
```ts
const data={"supervisorcomment":"OK"}
const response = await this.bpmnServer.engine.invoke({'items.id': taskId, 'items.status':'wait'}, data );
```
im use `"bpmn-server": "^1.4.6",`
![bugs](https://github.com/bpmnServer/bpmn-server/assets/1889979/5215766d-b8c1-4e0c-8dbe-9135b828a50a)

**Comment by ralphhanna** [comment]
What to do with user input?

not sure I understand, can repeat in a different phrasing

**Reply by kstan79** [reply]
Hm. I draw a flow to illustrate it later

**Reply by ralphhanna** [reply]
Sorry, in MongoDB item.data should not be there a bug!!!

**Reply by kstan79** [reply]
You mean in item should not have'data'?

Beside during invoke how to put actor input into vars?

**Comment by ralphhanna** [comment]
If variable name starts with vars. It Will be placed into vars

scripts can validate and move data between

this does not solve the problem of who has inputted the data?

Another question: should we allow users to input task data on ‘assign’?

**Reply by kstan79** [reply]
wow, I just know this. as long as data captured that it will be fine for me, I can set rule all property during invoke should start with 'vars', however will it still merge into instance `data`?

Next is comment regarding this ideal, prefix `vars` to determine it go in to 'vars` property is really not a good ideal. I was use similar dynamic approach in php, and it create lot of education and debug burden in long term. I feel it Is kind of 'no-no' in new programming world.

to develop and maintain good source code (especially cloud target platform), avoid special/hidden features is important and thats why bpmn-server support typescript isn't it?

in my opinion vars should have dedicated type and dedicated way to assing. change the invoke's data to `invoke vars` will be good approach

**Reply by kstan79** [reply]
here is sample user input which is less appropriate to go into `data`, thats the reason i keep emphasise don't go in by default, cause the data will be use by next process `officially`.

Since after 3 approval data going into accounting software (with final description copy over), it will print in `official invoice` and send `official email` to client.

Anyhow, I can accommadate with using `vars`, and remove `vars` property from `data` at the end. Just, I believe it is not good for bpmn-server in long term. 
```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn:definitions xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:camunda="http://camunda.org/schema/1.0/bpmn" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" id="Definitions_1" targetNamespace="http://bpmn.io/schema/bpmn">
  <bpmn:process id="Process_1" name="sample sales" isExecutable="false">
    <bpmn:startEvent id="StartEvent_1" name="new quotation">
      <bpmn:documentation>{
"date":"2000-01-01",
"preparedBy": "kstan",
"customer": "C1",
"total": 200.00,
"description": "ship via ABC123",
"details": [
{"product":"P1",qty:1,"unitprice":200},
{"product":"P2",qty:1,"unitprice":100}] 
}</bpmn:documentation>
      <bpmn:outgoing>Flow_0mt35re</bpmn:outgoing>
    </bpmn:startEvent>
    <bpmn:sequenceFlow id="Flow_0mt35re" sourceRef="StartEvent_1" targetRef="Activity_1gkqf1c" />
    <bpmn:userTask id="Activity_1gkqf1c" name="approve by supervisor" camunda:assignee="actor1">
      <bpmn:documentation>{
"action": "accept",
"description": "ok, look good"
}</bpmn:documentation>
      <bpmn:incoming>Flow_0mt35re</bpmn:incoming>
      <bpmn:outgoing>Flow_0ra567t</bpmn:outgoing>
    </bpmn:userTask>
    <bpmn:sequenceFlow id="Flow_0ra567t" sourceRef="Activity_1gkqf1c" targetRef="Gateway_0fz2vsa" />
    <bpmn:parallelGateway id="Gateway_0fz2vsa">
      <bpmn:incoming>Flow_0ra567t</bpmn:incoming>
      <bpmn:outgoing>Flow_1mkpjn1</bpmn:outgoing>
      <bpmn:outgoing>Flow_0ybeep7</bpmn:outgoing>
    </bpmn:parallelGateway>
    <bpmn:sequenceFlow id="Flow_1mkpjn1" sourceRef="Gateway_0fz2vsa" targetRef="Activity_0p5vqh6" />
    <bpmn:sequenceFlow id="Flow_0ybeep7" sourceRef="Gateway_0fz2vsa" targetRef="Activity_0mz5whr" />
    <bpmn:exclusiveGateway id="Gateway_0s86ww2">
      <bpmn:incoming>Flow_1mmex5f</bpmn:incoming>
      <bpmn:incoming>Flow_1ez6np4</bpmn:incoming>
      <bpmn:outgoing>Flow_0xr7pkc</bpmn:outgoing>
    </bpmn:exclusiveGateway>
    <bpmn:sequenceFlow id="Flow_1mmex5f" sourceRef="Activity_0mz5whr" targetRef="Gateway_0s86ww2" />
    <bpmn:sequenceFlow id="Flow_1ez6np4" sourceRef="Activity_0p5vqh6" targetRef="Gateway_0s86ww2" />
    <bpmn:sequenceFlow id="Flow_0xr7pkc" sourceRef="Gateway_0s86ww2" targetRef="Activity_1hpk5us" />
    <bpmn:serviceTask id="Activity_1hpk5us" name="duplicate pure Quotation data to invoice at accounting software">
      <bpmn:documentation>{
"date":"2000-01-01",
"preparedBy": "kstan",
"customer": "C1",
"total": 200.00,
"description": "ship via ABC123",
"details": [
{"product":"P1",qty:1,"unitprice":200},
{"product":"P2",qty:1,"unitprice":100}] 
}</bpmn:documentation>
      <bpmn:incoming>Flow_0xr7pkc</bpmn:incoming>
      <bpmn:outgoing>Flow_0f1zd6r</bpmn:outgoing>
    </bpmn:serviceTask>
    <bpmn:userTask id="Activity_0p5vqh6" name="branch manager approve" camunda:assignee="branchmanager1">
      <bpmn:documentation>{
"action": "approve",
"description": "ok we have sufficient stock, but near expired date. quickly sell it"
}</bpmn:documentation>
      <bpmn:incoming>Flow_1mkpjn1</bpmn:incoming>
      <bpmn:outgoing>Flow_1ez6np4</bpmn:outgoing>
    </bpmn:userTask>
    <bpmn:userTask id="Activity_0mz5whr" name="sales manager approve" camunda:assignee="salesmanager1">
      <bpmn:documentation>{
"action": "accept",
"description": "ok, I'm good with this client, this quotation we have good margin"
}</bpmn:documentation>
      <bpmn:incoming>Flow_0ybeep7</bpmn:incoming>
      <bpmn:outgoing>Flow_1mmex5f</bpmn:outgoing>
    </bpmn:userTask>
    <bpmn:sequenceFlow id="Flow_0f1zd6r" sourceRef="Activity_1hpk5us" targetRef="Event_0qgn5ud" />
    <bpmn:endEvent id="Event_0qgn5ud" name="send pure invoice details client">
      <bpmn:documentation>{
"date":"2000-01-01",
"preparedBy": "kstan",
"customer": "C1",
"total": 200.00,
"description": "ship via ABC123",
"details": [
{"product":"P1",qty:1,"unitprice":200},
{"product":"P2",qty:1,"unitprice":100}] 
}</bpmn:documentation>
      <bpmn:incoming>Flow_0f1zd6r</bpmn:incoming>
      <bpmn:messageEventDefinition id="MessageEventDefinition_0v5juel" />
    </bpmn:endEvent>
  </bpmn:process>
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="Process_1">
      <bpmndi:BPMNShape id="_BPMNShape_StartEvent_2" bpmnElement="StartEvent_1">
        <dc:Bounds x="132" y="122" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="116" y="165" width="69" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_06csff4_di" bpmnElement="Activity_1gkqf1c">
        <dc:Bounds x="220" y="100" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Gateway_1dgtb0r_di" bpmnElement="Gateway_0fz2vsa">
        <dc:Bounds x="375" y="115" width="50" height="50" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Gateway_0s86ww2_di" bpmnElement="Gateway_0s86ww2" isMarkerVisible="true">
        <dc:Bounds x="635" y="225" width="50" height="50" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_060fdea_di" bpmnElement="Activity_1hpk5us">
        <dc:Bounds x="740" y="210" width="100" height="80" />
        <bpmndi:BPMNLabel />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_1v5i0b9_di" bpmnElement="Activity_0p5vqh6">
        <dc:Bounds x="480" y="100" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_10bbwzn_di" bpmnElement="Activity_0mz5whr">
        <dc:Bounds x="480" y="210" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_1sah643_di" bpmnElement="Event_0qgn5ud">
        <dc:Bounds x="902" y="232" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="877" y="275" width="87" height="27" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="Flow_0mt35re_di" bpmnElement="Flow_0mt35re">
        <di:waypoint x="168" y="140" />
        <di:waypoint x="220" y="140" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0ra567t_di" bpmnElement="Flow_0ra567t">
        <di:waypoint x="320" y="140" />
        <di:waypoint x="375" y="140" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1mkpjn1_di" bpmnElement="Flow_1mkpjn1">
        <di:waypoint x="425" y="140" />
        <di:waypoint x="480" y="140" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0ybeep7_di" bpmnElement="Flow_0ybeep7">
        <di:waypoint x="400" y="165" />
        <di:waypoint x="400" y="250" />
        <di:waypoint x="480" y="250" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1mmex5f_di" bpmnElement="Flow_1mmex5f">
        <di:waypoint x="580" y="250" />
        <di:waypoint x="635" y="250" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1ez6np4_di" bpmnElement="Flow_1ez6np4">
        <di:waypoint x="580" y="140" />
        <di:waypoint x="660" y="140" />
        <di:waypoint x="660" y="225" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0xr7pkc_di" bpmnElement="Flow_0xr7pkc">
        <di:waypoint x="685" y="250" />
        <di:waypoint x="740" y="250" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0f1zd6r_di" bpmnElement="Flow_0f1zd6r">
        <di:waypoint x="840" y="250" />
        <di:waypoint x="902" y="250" />
      </bpmndi:BPMNEdge>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn:definitions>
```

**Comment by ralphhanna** [comment]
To answer your question; the input variables that start with 'vars' will only be saved in `item.vars` not in `instance.data` other variables will be saved in `instance.data`.

Also keep in mind, that both 'item.vars` and 'instance.data` can contain other objects so you can qualify the data with **Official** once accepted.

**Reply by kstan79** [reply]
> To answer your question; the input variables that start with 'vars' will only be saved in item.vars not in instance.data other variables will be saved in instance.data.

this is perfect!

> Also keep in mind, that both 'item.vars and 'instance.data can contain other objects so you can qualify the data with Official once accepted.

agree. you are the person who make final decision and I have to obey it. I just highlight base on some experience, which enhancement I feel is healthier for project. As previous post I can accommodate since I can obtain value from `vars`.

**Reply by kstan79** [reply]
hm.. no it "varssample" still not go in `vars` and it merged into `data`:

```ts
const response = await this.bpmnServer.engine.invoke({'items.id': taskId, 'items.status':'wait'}, {varssample:"test vars"} );
```

![bugs2](https://github.com/bpmnServer/bpmn-server/assets/1889979/3e1090f1-3bf3-45c2-8dd1-97e37e9850fc)

**Comment by ralphhanna** [comment]
Sorry, not yet implemented, will be in next release.

Thanks
________________________________

**Comment by ralphhanna** [comment]
One other option to consider for the above issues for `User Task` is as follows 
1. input variables and values are passed to `item.input` not directly into `instance.data`
2. A Script on `Start` Event; can move data as you wish to item.vars and remove them from item.input 
3. system will by default copy contents of item.input (what is left of it) into instance.data

P.S. `item.input` and `item.output` are not saved in DB

Your feedback please?

**Reply by kstan79** [reply]
ya input/output is the correct way of manage input and output on every single node in bpmn. I'm feel glade of you decided to support input/output variables. However, using input/output in bpmn start to make complexity increase so much. 

Generally `I not that recommend` you implement input/output variables at the time being. Cause:
1. there is a lot of decision you have to make (issue is not only programming), a lot scenario you have to test, such as:
   a. where to obtain inputs from user task
   b. where to obtain inpus from service task
   c. the data types of inputs and outputs, where to store the schema
   d. how to prepare single or multiple output data in service task
   e. how to prepare single or multiple output data in user task
   f. you will notice camunda modeler provide only simple input/output variables UI. It can't define data type (boolean, interface), UI taken so much spaces. There is no `initial` variables definition in editor too
   g. event there is perfect UI editor, audit of every single input/output vars is tedious, and easily map wrongly.

2. overall, input/output variables is must have features, but it is big `pandora box` for you at the moment.  You may delay this implementation since:
  a. currently camunda modeler is not perfect for input/output variable, no point bother it.
  b. previous discussion in `invoke` is talking about `single field input vars` in usertask. which can serve most of the purpose, complex scenario we can let developer manage it inside deligates.
  c. You may don't want mess up your work, cause you have to change the execution of every nodes which consider stable. example service task delegates:
```ts

//3 inputs in service task, will have 3 input in delegate service
hello(recordId: string, customer: customerObject1, totalAmount: number ){
   //do something here

   // user define 2 output variables (customer, messageStatus), but here you only can return 1, you start headache how to handle 2nd vars
   return customer
}
```
 d. after evaluate `bpmn-server`, no implement input/output variables is ok. We can use `data`, `vars` to handle general task.
 e. there is still plenty of place in `bpmn-server` need polish, let focus to make it good for production use and grow community. 

I saw your #142, and I feel you shall follow your standardization, which is:
- `$(this.data)`
- or `${this.data}`
- 1 (no quote) as number
- "1" (with quote) as string

**Comment by ralphhanna** [comment]
Thanks for your valuable feedback, always appreciated, keep it coming.

1. bpmn-server always supported input/output [see]( https://github.com/bpmnServer/bpmn-server/blob/master/docs/data.md)
2. I tried to keep it as simple as possible

I agree, managing Input/Output for User Tasks does not make a lot of sense, so rolling back this idea, Thanks.

I like what you said about :
```js
$(this.data)
or ${this.data}
1 (no quote) as number
"1" (with quote) as string
```
But it breaks older releases ,the old syntax is used for:
1. Conditional Flow
2. Loops
3. I/O parameter
Will have to write a script to warn developers to upgrade and show invalid usage

**Reply by kstan79** [reply]
You have to apply single rules for every place, that can avoid confusion and make your regression reusable.
